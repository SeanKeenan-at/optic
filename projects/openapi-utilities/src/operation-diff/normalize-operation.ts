import { OpenAPIV3 } from 'openapi-types';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { Coordinates } from './coordinates';
import JsonPointerHelpers from '@useoptic/json-pointer-helpers/build/json-pointers/json-pointer-helpers';

export function normalizeOperation(
  spec: OpenAPIV3.Document,
  method: OpenAPIV3.HttpMethods,
  path: string
): OpenAPIOperation {
  const coordinateAliases: {
    [key: string]: string;
  } = {};

  const operationCoordinate = jsonPointerHelpers.compile([
    'paths',
    path,
    method,
  ]);

  let operation = jsonPointerHelpers.get(spec, operationCoordinate);

  // in-source security
  if (spec.security && !operation.security) {
    operation = { ...operation, security: spec.security };
    coordinateAliases[
      JsonPointerHelpers.append(operationCoordinate, 'security')
    ] = JsonPointerHelpers.compile(['security']);
  }

  if (spec.paths[path]!.parameters) {
    const sharedParams = spec.paths[path]!.parameters || [];
    const operationParamsCount = (operation.parameters || []).length;
    operation = {
      ...operation,
      parameters: [...(operation.parameters || []), ...sharedParams],
    };
    sharedParams.forEach((sharedParameter, index) => {
      coordinateAliases[
        JsonPointerHelpers.append(
          operationCoordinate,
          'parameters',
          (operationParamsCount + index).toString()
        )
      ] = jsonPointerHelpers.compile([
        'paths',
        path,
        'parameters',
        index.toString(),
      ]);
    });
  }

  return {
    method,
    path,
    context: {
      environments: spec.servers || [],
      securitySchemas: spec?.components?.securitySchemes as any, // need to fork openapi-type and get rid of $ref
      sharedParameters: spec.paths[path]!.parameters || ([] as any), // need to fork openapi-type and get rid of $ref,
    },
    operation,
    coordinates: {
      absoluteJsonPath: jsonPointerHelpers.compile(['paths', path, method]),
    },
    coordinateAliases,
  };
}

export interface OpenAPIOperation {
  method: string;
  path: string;
  context: {
    environments: OpenAPIV3.ServerObject[];
    securitySchemas: {
      [key: string]: OpenAPIV3.SecuritySchemeObject;
    };
    sharedParameters: OpenAPIV3.ParameterObject[];
  };
  operation: OpenAPIV3.OperationObject;
  coordinates: Coordinates;
  // for post-processing rule results
  coordinateAliases: {
    [key: string]: string;
  };
}
