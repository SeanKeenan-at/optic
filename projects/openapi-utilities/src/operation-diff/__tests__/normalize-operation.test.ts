import fs from 'fs-extra';
import { normalizeOperation } from '../normalize-operation';
import { OpenAPIV3 } from 'openapi-types';

const jsonFromFile = async (path: string) => {
  const bytes = await fs.readJson(path);
  return bytes;
};

it('can normalize an openapi operation', async () => {
  const spec = await jsonFromFile(
    './inputs/openapi3/petstore0.json.flattened-without-sourcemap.json'
  );

  const operation = normalizeOperation(
    spec,
    OpenAPIV3.HttpMethods.POST,
    '/pet/{petId}/uploadImage'
  );

  expect(operation).toMatchSnapshot();
});
