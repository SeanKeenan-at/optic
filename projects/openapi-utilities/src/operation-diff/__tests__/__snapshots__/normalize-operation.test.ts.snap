// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can normalize an openapi operation 1`] = `
Object {
  "context": Object {
    "environments": Array [
      Object {
        "url": "https://petstore.swagger.io/v2",
      },
      Object {
        "url": "http://petstore.swagger.io/v2",
      },
    ],
    "securitySchemas": Object {
      "api_key": Object {
        "in": "header",
        "name": "api_key",
        "type": "apiKey",
      },
      "petstore_auth": Object {
        "flows": Object {
          "implicit": Object {
            "authorizationUrl": "http://petstore.swagger.io/oauth/dialog",
            "scopes": Object {
              "read:pets": "read your pets",
              "write:pets": "modify pets in your account",
            },
          },
        },
        "type": "oauth2",
      },
    },
    "sharedParameters": Array [],
  },
  "coordinateAliases": Object {},
  "coordinates": Object {
    "absoluteJsonPath": "/paths/~1pet~1{petId}~1uploadImage/post",
  },
  "method": "post",
  "operation": Object {
    "operationId": "uploadFile",
    "parameters": Array [
      Object {
        "description": "ID of pet to update",
        "in": "path",
        "name": "petId",
        "required": true,
        "schema": Object {
          "format": "int64",
          "type": "integer",
        },
      },
      Object {
        "description": "A debug token",
        "in": "cookie",
        "name": "debug",
        "schema": Object {
          "type": "integer",
        },
      },
    ],
    "requestBody": Object {
      "content": Object {
        "multipart/form-data": Object {
          "schema": Object {
            "properties": Object {
              "additionalMetadata": Object {
                "description": "Additional data to pass to server",
                "type": "string",
              },
              "file": Object {
                "description": "file to upload",
                "format": "binary",
                "type": "string",
              },
            },
          },
        },
      },
    },
    "responses": Object {
      "200": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "code": Object {
                  "format": "int32",
                  "type": "integer",
                },
                "message": Object {
                  "type": "string",
                },
                "type": Object {
                  "type": "string",
                },
              },
              "type": "object",
            },
          },
        },
        "description": "successful operation",
      },
    },
    "security": Array [
      Object {
        "petstore_auth": Array [
          "write:pets",
          "read:pets",
        ],
      },
    ],
    "summary": "uploads an image",
    "tags": Array [
      "pet",
    ],
  },
  "path": "/pet/{petId}/uploadImage",
}
`;
